src/gazebo_mavlink_interface.cpp

```diff
@@ -615,16 +657,40 @@ void GazeboMavlinkInterface::send_mavlink_message(const mavlink_message_t *messa
     uint8_t buffer[MAVLINK_MAX_PACKET_LEN];
     int packetlen = mavlink_msg_to_send_buffer(buffer, message);
 
-    ssize_t len;
-    if (use_tcp_) {
-      len = send(simulator_tcp_client_fd_, buffer, packetlen, 0);
-    } else {
-      len = sendto(simulator_socket_fd_, buffer, packetlen, 0, (struct sockaddr *)&remote_simulator_addr_, remote_simulator_addr_len_);
-    }
+    if (fds_[CONNECTION_FD].fd > 0) {
+      int timeout_ms = (received_first_actuator_ && enable_lockstep_) ? 1000 : 0;
+      int ret = ::poll(&fds_[0], N_FDS, timeout_ms);
 
-    if (len <= 0)
-    {
-      gzerr << "Failed sending mavlink message: " << strerror(errno) << "\n";
+      if (ret < 0) {
+        gzerr << "poll error: " << strerror(errno) << "\n";
+        return;
+      }
+
+      if (ret == 0 && timeout_ms > 0) {
+        gzerr << "poll timeout\n";
+        return;
+      }
+
+      if (!(fds_[CONNECTION_FD].revents & POLLOUT)) {
```

Comment:
Can't you just initialize without `POLLOUT` and then this will never happen?

Reply:
Hmm, the connection accepted at the socket with id at CONNECTION_FD is both read/write. We need POLLOUT for sending data to PX4. 

The reason I used polling for sending data (just like receiving data) is that so we can check whether calling send() is possible before calling it. Otherwise, sudden disconnections (Ctrl+C) from the client made the socket get stuck in TIME_WAIT on calling send() on a broken connection. This didn't allow reconnection right away.